libs/librepcb/common/widgets/tabwidget.cpp

```diff
@@ -0,0 +1,61 @@
+/*
+ * LibrePCB - Professional EDA for everyone!
+ * Copyright (C) 2013 LibrePCB Developers, see AUTHORS.md for contributors.
+ * https://librepcb.org/
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*******************************************************************************
+ *  Includes
+ ******************************************************************************/
+#include "tabwidget.h"
+
+#include <QtCore>
+#include <QtWidgets>
+
+/*******************************************************************************
+ *  Namespace
+ ******************************************************************************/
+namespace librepcb {
+
+TabWidget::TabWidget(QWidget* parent) : QTabWidget{parent} {
+  tabBar()->installEventFilter(this);
+}
+
+bool TabWidget::eventFilter(QObject* o, QEvent* e) {
+  // Handle middle mouse click on closable tabs
+  bool isTabBar           = o == tabBar();
+  bool isMouseButtonPress = e->type() == QEvent::MouseButtonPress;
+  bool tabsAreClosable    = tabBar()->tabsClosable();
+  if (isTabBar && isMouseButtonPress && tabsAreClosable) {
+    auto me = static_cast<QMouseEvent*>(e);
+    if (me->buttons() == Qt::MiddleButton) {
+      auto tabIdx = tabBar()->tabAt(me->pos());
+      if (tabIdx >= 0 &&
+          tabBar()->tabButton(tabIdx, QTabBar::RightSide) != nullptr) {
```

Comment:
I would remove this check because the right side button might not always be the close button. See [Qt docs](http://doc.qt.io/qt-5/qtabbar.html#tabsClosable-prop):

> When tabsClosable is set to true a close button will appear on the tab on either the left or right hand side depending upon the style. When the button is clicked the tab the signal tabCloseRequested will be emitted.

In addition, there might be other ways to close a tab, for example a context menu.

But that's fine, it's up to the receiver of the `tabCloseRequested()` signal anyway to actually close the tab. So if you emit that signal on a non-closable tab the receiver can just ignore it. In our case this check should be moved to `LibraryEditor::tabCloseRequested()` instead.